cirq/ops/parity_gates_test.py

```diff
@@ -0,0 +1,322 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for `parity_gates.py`."""
+
+import numpy as np
+
+import cirq
+
+
+def test_xx_init():
+    assert cirq.XXPowGate(exponent=1).exponent == 1
+    v = cirq.XXPowGate(exponent=0.5)
+    assert v.exponent == 0.5
+
+
+def test_xx_eq():
+    eq = cirq.testing.EqualsTester()
+    eq.add_equality_group(cirq.XX,
+                          cirq.XXPowGate(),
+                          cirq.XXPowGate(exponent=1, global_shift=0),
+                          cirq.XXPowGate(exponent=3, global_shift=0))
+    eq.add_equality_group(cirq.XX**0.5, cirq.XX**2.5, cirq.XX**4.5)
+    eq.add_equality_group(cirq.XX**0.25, cirq.XX**2.25, cirq.XX**-1.75)
+
+    iXX = cirq.XXPowGate(global_shift=0.5)
+    eq.add_equality_group(iXX**0.5, iXX**4.5)
+    eq.add_equality_group(iXX**2.5, iXX**6.5)
+
+
+def test_xx_pow():
+    assert cirq.XX**0.5 != cirq.XX**-0.5
+    assert cirq.XX**-1 == cirq.XX
+    assert (cirq.XX**-1)**0.5 == cirq.XX**-0.5
+
+
+def test_xx_str():
+    assert str(cirq.XX) == 'XX'
+    assert str(cirq.XX**0.5) == 'XX**0.5'
+    assert str(cirq.XXPowGate(global_shift=0.1)) == 'XX'
+
+    ms = cirq.XXPowGate(global_shift=-0.5)
+    assert str(ms) == 'MS(π/2)'
+    assert str(ms**2) == 'MS(2.0π/2)'
+    assert str(ms**-1) == 'MS(-1.0π/2)'
+
+
+def test_xx_repr():
+    assert repr(cirq.XXPowGate()) == 'cirq.XX'
+    assert repr(cirq.XXPowGate(exponent=0.5)) == '(cirq.XX**0.5)'
+
+    cirq.testing.assert_equivalent_repr(cirq.XX)
+    cirq.testing.assert_equivalent_repr(cirq.XX**0.1)
+    cirq.testing.assert_equivalent_repr(cirq.XX**0.5)
+    cirq.testing.assert_equivalent_repr(cirq.XX**2)
+    cirq.testing.assert_equivalent_repr(cirq.XXPowGate(exponent=0.2,
+                                                       global_shift=0.3))
+
+    ms = cirq.XXPowGate(global_shift=-0.5)
+    assert (repr(ms) == 'cirq.MS(np.pi/2)')
+    assert (repr(ms**2) == 'cirq.MS(2.0*np.pi/2)')
+    assert (repr(ms**-0.5) == 'cirq.MS(-0.5*np.pi/2)')
+
+    cirq.testing.assert_equivalent_repr(ms)
+    cirq.testing.assert_equivalent_repr(ms**0.1)
+    cirq.testing.assert_equivalent_repr(ms**0.5)
+    cirq.testing.assert_equivalent_repr(ms**2)
+
+
+def test_xx_matrix():
+    np.testing.assert_allclose(cirq.unitary(cirq.XX),
+                               np.array([[0, 0, 0, 1],
+                                         [0, 0, 1, 0],
+                                         [0, 1, 0, 0],
+                                         [1, 0, 0, 0]]),
+                               atol=1e-8)
+    np.testing.assert_allclose(cirq.unitary(cirq.XX**2),
+                               np.eye(4),
+                               atol=1e-8)
+    c = np.cos(np.pi/4)
```

Comment:
You'd cover more possible mistakes with other angles since sin(π/4) == cos(π/4). An easy choice is something like π/8 which corresponds to an exponent that may be represented exactly, but I recommend first checking whether errors with an angle like π/7 still fit within the tolerance (and if so, let's use π/7 or similar).

Reply:
Done.